src/stdlib.ts

```
@@ -1,17 +1,1788 @@
+            const result: bigint = intVal < 0 ? -intVal : intVal;
+            return { type: 'int', value: result };
+        }
+        case 'number': {
+            return { type: 'number', value: Math.abs(x.value) };
+        }
+        case 'complex': {
+            // Calculate the modulus (absolute value) of a complex number.
+            const real = x.value.real;
+            const imag = x.value.imag;
+            const modulus = Math.sqrt(real * real + imag * imag);
+            return { type: 'number', value: modulus };
+        }
+        default:
+            handleRuntimeError(context, new TypeError(source, command as es.Node, context, args[0].type, "float', 'int' or 'complex"));
+    }
+}
+
+function toStr(val: Value): string {
+    return String(val.value);
+}
+
+export function error(args: Value[], source: string, command: ControlItem, context: Context): Value {
+    const output = "Error: " + args.map(arg => toStr(arg)).join(' ') + '\n';
+    throw new Error(output);
+}
+
+export function isinstance(args: Value[], source: string, command: ControlItem, context: Context): Value {
+    if (args.length < 2) {
+        handleRuntimeError(context, new MissingRequiredPositionalError(source, (command as es.Node), 'isinstance', Number(2), args, false));
+    } else if (args.length > 2) {
+        handleRuntimeError(context, new TooManyPositionalArgumentsError(source, (command as es.Node), 'isinstance', Number(2), args, false));
+    }
+  
+    const obj = args[0];
+    const classinfo = args[1];
+
+    let expectedType: string;
+    if (classinfo.type === 'string') {
+        switch (classinfo.value) {
+            case 'int':
+                expectedType = 'bigint';
+                break;
+            case 'float':
+                expectedType = 'number';
+                break;
+            case 'string':
+                expectedType = 'string';
+                break;
+            case 'bool':
+                expectedType = 'bool';
```

`isinstance(True, int)` is true in Python (bool is a subclass of int in CPython). Do we want to allow that?

---

No... we follow Source Chapter 1

---

Hmm, our current definition of "sublanguage" is like this:

A program that runs successfully using Python §1 in Source Academy should run successfully in the official Python implementation and produce the same result. (assuming that the correct sicp library is loaded that provides things like Math_sin)

By that reasoning, if we include isinstance in Python §1, any call of isinstance that succeeds in Python §1 should give the same result.

So currently, does isinstance(True, int) give an error, or does it give False in Python §1? An error would be consistent with our sublanguage notion, but False would not be.

---

> Hmm, our current definition of "sublanguage" is like this:
> 
> A program that runs successfully using Python §1 in Source Academy should run successfully in the official Python implementation and produce the same result. (assuming that the correct sicp library is loaded that provides things like Math_sin)
> 
> By that reasoning, if we include isinstance in Python §1, any call of isinstance that succeeds in Python §1 should give the same result.
> 
> So currently, does isinstance(True, int) give an error, or does it give False in Python §1? An error would be consistent with our sublanguage notion, but False would not be.

Thanks for the clarification. I will updated the implementation so that isinstance(True, int) now raises an error instead of returning False, to stay consistent with the sublanguage design.